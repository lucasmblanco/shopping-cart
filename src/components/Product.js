import React, { useEffect } from 'react';
import { useState } from 'react';

export default function Product({ data, addProduct, quantity, status }) {
  const [addedStatus, setAddedStatus] = useState(status);
  const itemQuantity = quantity;

  useEffect(() => {
    setAddedStatus(status);
  }, [status]);

  return (
    <div
      data-testid="shop-product"
      className=" h-product-cards bg-white rounded-[calc(0.5rem_+_2px)] flex flex-col items-center shadow-xl border-solid border-2
     border-light-red-accent animate-fade-in-bck-open">
      <div className="w-80 h-56 p-5 flex justify-center">
        <img className="h-full " src={data.image} alt={data.title} />
      </div>
      <div className=" flex justify-between flex-col basis-full items-center rounded-br-lg text-white rounded-bl-lg bg-red-accent h-full w-full text-center">
        <div className=" text-sm text-center p-3 w-full ">{data.title}</div>
        <div className="text-gray-600 text-xs w-full">
          {data.category.toUpperCase()}
        </div>
        <div className=" text-xs w-full">{data.rating.rate + '/5'}</div>
        <div className=" text-2xl w-full m-none bg-red-500 p-3 tracking-wide">
          {'$' + data.price}
        </div>

        {addedStatus ? (
          <div
            data-testid="product-added-message"
            className="text-lg bg-red-accent p-4 m-3 border-2 border-transparent">
            IN THE CART
          </div>
        ) : (
          <button
            className="text-lg bg-red-accent p-4 m-3 rounded-xl hover:bg-light-red-accent
       border-light-red-accent hover:border-red-accent border-2 transition-all duration-500"
            onClick={() =>
              addProduct({
                ...data,
                added: addedStatus,
                quantity: itemQuantity,
              })
            }>
            ADD TO CART
          </button>
        )}
      </div>
    </div>
  );
}
