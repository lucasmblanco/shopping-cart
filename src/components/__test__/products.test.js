import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event';
import Product from '../Product';

const mockProduct = {
  id: 1,
  title: 'MockProduct',
  image: 'abc.jpg',
  rating: {
    rate: 1,
  },
  price: 1,
  category: 'MockCategory',
};

const ADD_MOCK = jest.fn();

const STATUS_MOCK = true;

it('should render the component with passing data', () => {
  const { asFragment } = render(<Product data={mockProduct} />);

  expect(asFragment()).toMatchSnapshot();
});

it('should add product to cart', async () => {
  render(<Product data={mockProduct} addProduct={ADD_MOCK} />);

  const addButton = screen.getByText('ADD TO CART');

  await userEvent.click(addButton);

  expect(ADD_MOCK).toHaveBeenCalled();
});

it('should show product status', async () => {
  render(
    <Product data={mockProduct} addProduct={ADD_MOCK} status={STATUS_MOCK} />,
  );

  const status = screen.getByTestId('product-added-message');

  expect(status).toBeInTheDocument();
});
