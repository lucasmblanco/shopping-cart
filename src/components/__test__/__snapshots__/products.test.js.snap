// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the component with passing data 1`] = `
<DocumentFragment>
  <div
    class=" h-product-cards bg-white rounded-[calc(0.5rem_+_2px)] flex flex-col items-center shadow-xl border-solid border-2
 border-light-red-accent animate-fade-in-bck-open"
    data-testid="shop-product"
  >
    <div
      class="w-80 h-56 p-5 flex justify-center"
    >
      <img
        alt="MockProduct"
        class="h-full "
        src="abc.jpg"
      />
    </div>
    <div
      class=" flex justify-between flex-col basis-full items-center rounded-br-lg text-white rounded-bl-lg bg-red-accent h-full w-full text-center"
    >
      <div
        class=" text-sm text-center p-3 w-full "
      >
        MockProduct
      </div>
      <div
        class="text-gray-600 text-xs w-full"
      >
        MOCKCATEGORY
      </div>
      <div
        class=" text-xs w-full"
      >
        1/5
      </div>
      <div
        class=" text-2xl w-full m-none bg-red-500 p-3 tracking-wide"
      >
        $1
      </div>
      <button
        class="text-lg bg-red-accent p-4 m-3 rounded-xl hover:bg-light-red-accent
 border-light-red-accent hover:border-red-accent border-2 transition-all duration-500"
      >
        ADD TO CART
      </button>
    </div>
  </div>
</DocumentFragment>
`;
